Enable extra addons from all-packages media

QingmingSu opened the pull request:
In upgrade testing, the sle addons, including extensions and
modules, are defined with SCC_ADDONS. During system upgrade
with all-packages media, the addons installed on the original
system should be mapped to new ones provided by the media.
- see poo#27504 for details

- Related ticket: https://progress.opensuse.org/issues/27504
- Needles: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/628
- Verification run: 
        [migration_offline_media_sles12sp3_ha+geo+sdk+we]
        http://openqa-apac1.suse.de/tests/29
        The test failure in releasenotes looks like a product bug.
        [migration_offline_media_sles12sp3_asmm+contm+hpcm+lgm+pcm+tcm+wsm]
        http://openqa-apac1.suse.de/tests/30

QingmingSu commented:
WIP: Need more adaption and verification.

ldevulder commented on tests/installation/addon_products_sle.pm:
```diff
@@ -22,18 +22,61 @@ use registration '%SLE15_DEFAULT_MODULES';
+    # wsm -- script
+    # Not good idea to use different module names, we have to bridge the
+    # gap here to use existing needles
+    my %addons_name_map = (
```
Looks goot for SLE15, but in SLE12-SP4 SLE-HA-GEO is still in a separate extension, so is this modification will not break SLE12-SPn to SLE12-SPn+1 upgrade?

QingmingSu replied:
Hmm, it will break SLE12-SPn to SLE12-SPn+1 upgrade. The rename should be restricted to sle15 or later.
Thanks for the comment.

QingmingSu replied:
Not likely to happen, since this piece of test code only apply for all-packages media but SLE12-SP4 should not have such addon media.
Anyway, I will restrict the scc addon renaming to sle15 or later.

QingmingSu commented:
Updated: 
- move scc addons renaming to lib/registration.pm as function rename_scc_addons
- restrict the scc addons renaming to sle15 or later

Verified in new test runs:
[migration_offline_media_sles12sp3_ha+geo+sdk+we]
http://openqa-apac1.suse.de/tests/31
The test failure in releasenotes looks like a product bug.
[migration_offline_media_sles12sp3_asmm+contm+hpcm+lgm+pcm+tcm+wsm]
http://openqa-apac1.suse.de/tests/32

ldevulder commented on lib/registration.pm:
```diff
@@ -509,5 +515,35 @@ sub scc_deregistration {
+# Not good idea to use different module names, we have to bridge the
+# gap here to use existing needles
+sub rename_scc_addons {
+    return unless get_var('SCC_ADDONS') && sle_version_at_least('15');
```
For testing SLE version you also need to use `is_sle` like this: `is_sle && sle_version_...`.

QingmingSu replied:
Maybe we can just ignore the is_sle here? It is unlikely to define SCC_ADDONS openSUSE testing.

ldevulder replied:
Yes and no :-) Because `sle_version_at_least` can't just be use alone to test SLE OS. It's not just me (I had the same comments on my PR), it's just "needed".

I plan to try to modify this in the future, I need to open a discussion on poo# and/or #testing for this.

okurz replied:
Actually here we can leave without it. `sle_version_at_least` does not "need" `is_sle` it is rather that `sle_version_at_least` might not behave as one would expect on openSUSE. But for now openSUSE does not use `lib/registration.pm` at all so we should not need to care that much.

ldevulder commented:
The failure in releasenotes seems to be due to a missing needle for SLE-HA extension, maybe you could try to add it :-)

QingmingSu commented:
@ldevulder Yes the failure in releasenotes actually indicates two issues:
1, Test issue that addon release notes, like SLE-HA, are not checked, plan to fix the issue in another ticket.
2, Product issue that SLE15 release notes is not shown, here we expect both SLE 15 release notes and addon's release notes.

okurz commented on tests/installation/addon_products_sle.pm:
```diff
@@ -17,23 +17,49 @@ use testapi;
     push @addons, 'desktop' if check_var('DESKTOP', 'gnome') && !grep(/^desktop$/, @addons);
+
+    # The legacy module is required if upgrade from previous version
+    push @addons, 'legacy' if get_var('UPGRADE') && !grep(/^legacy$/, @addons);
```
hm, why is that? I don't think it's the idea of the legacy module to be always required

QingmingSu replied:
The legacy module turns out to be always required in upgrading, which is noted in https://bugzilla.suse.com/show_bug.cgi?id=1066338#c5
It says:
"If you update from "pain" SLES12 without modules added to SLES15, you need minimal the basesystem module, the server application module, the desktop application module and the legacy module. Everything else will not work."

I think such info should be included in release notes or migration guide (if we have such doc).

okurz replied:
I guess for the test it is fine if you just add "(bsc#1066338)" to the comment in L31. That would help to find the comment by tkukuk and therefore the reasoning from development point of view.

okurz commented on tests/installation/addon_products_sle.pm:
```diff
@@ -17,23 +17,49 @@ use testapi;
+    # In upgrade testing, the sle addons, including extensions and modules,
+    # are defined with SCC_ADDONS, thus the addons could be patched on
+    # the original system (the system-to-be-upgraded).
+    # During system upgarde with all-packages media, the addons installed
```
s/upgarde/upgrade/

QingmingSu replied:
Typo fixed.

QingmingSu commented:
Updated %SLE15_MODULES in lib/registration.pm to define "contm" for Containers module and "pcm" for Public-Cloud module. Also redefine 'script' as Web-Scripting.

The "contm" and "pcm" have already been used in SCC_ADDONS, we should avoid different names for the same module (in case someone try to define new names for SLE15).

Also, needles have been created with names "contm" and "pcm", we have to create new needles with new names.

okurz commented:
s/upgarde/upgrade/ in commit message and PR description

QingmingSu commented:
@okurz Thanks for your comments. So carefully you checked.
I have added bug ID to comment line (L31) and fixed 'upgrade' typo both in commit message and PR description.

okurz commented:
the needles PR in PR description looks invalid: Needles: https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/628
Better remove that or replace by according gitlab.suse.de MR. Rest LGTM

QingmingSu commented:
The needles are critical against this PR, so we have to wait until the needles MR (628) get reviewed and merged.

QingmingSu commented:
The needle MR has been merged (https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/628)
